src/Microsoft.DotNet.ProjectModel/ProjectContext.cs

```diff
@@ -154,6 +155,100 @@ public static IEnumerable<ProjectContext> CreateContextForEachTarget(string proj
                         .BuildAllTargets();
         }
 
+        public static bool IsRidCompatibleWith(string rid, string otherRid)
+        {
+            if (otherRid == null)
+            {
+                return false;
+            }
+
+            if (rid == otherRid)
+            {
+                return true;
+            }
+
+            if (rid.EndsWith("x86") != otherRid.EndsWith("x86"))
+            {
+                return false;
+            }
+
+            if (rid.StartsWith("win10-"))
+            {
+                return otherRid.StartsWith("win8-") || otherRid.StartsWith("win7-");
+            }
+
+            if (rid.StartsWith("win8-"))
+            {
+                return otherRid.StartsWith("win7-");
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Creates a project context based on current platform
+        /// </summary>
+        /// <param name="buildConfiguration">build configuration</param>
+        /// <returns></returns>
+
+        public ProjectContext CreateRuntimeContextForCurrentPlatform(string buildConfiguration)
+        {
+            if (IsPortable)
+            {
+                return this;
+            }
+
+            string currentRid = PlatformServices.Default.Runtime.GetRuntimeIdentifier();
+
+            List<ProjectContext> contexts = CreateAllRuntimeContexts(buildConfiguration).ToList();
+            foreach (var runtimeContext in contexts)
+            {
+                if (runtimeContext.RuntimeIdentifier == currentRid)
+                {
+                    return runtimeContext;
+                }
+            }
+
+            foreach (var runtimeContext in contexts)
+            {
+                if (IsRidCompatibleWith(currentRid, runtimeContext.RuntimeIdentifier))
+                {
+                    return runtimeContext;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Creates a list of project contexts representing runtime based on a given context
+        /// </summary>
+        /// <param name="buildConfiguration">build configuration</param>
+        /// <returns>List of project contexts - one for each runtime identifier and one for runtime independent IL code (RuntimeIdentifier == null)</returns>
+
+        public IEnumerable<ProjectContext> CreateAllRuntimeContexts(string buildConfiguration)
+        {
+            if (!ProjectFile.HasRuntimeOutput(buildConfiguration))
+            {
+                return new[] { this };
+            }
+
+            IEnumerable<ProjectContext> allContexts = CreateContextForEachTarget(ProjectFile.ProjectFilePath);
```

Comment:
There's this method for creating / caching runtime contexts: [https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.ProjectModel/WorkspaceContext.cs#L187](https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.ProjectModel/WorkspaceContext.cs#L187)

I'd change that factory method to return a collection rather than (if I'm reading it right) the first runtime it finds and delete the one in ProjectContext (and also the one in DotnetProjectBuilder)

@anurse, does that sound right?

---
In reply to: [61349348](https://github.com/dotnet/cli/pull/2631#discussion_r61349348) [](ancestors = 61349348)